frontend/src/App.tsx

```diff
@@ -454,6 +460,14 @@ export class App extends PureComponent<Props, State> {
     }
   }
 
+  /**
+   * Handler for ForwardMsg.newUrl messages
+   * @param newUrl string
+   */
+  handleNewUrlChange = (newUrl: string): void => {
+    window.history.pushState({}, "", newUrl ? newUrl : "/")
```

Comment:
this has been changed to append queryString using `window.history.pushState` such the browser will not be redirected or refreshed. 

this will only be called upon user calls `st.experimental.set_query_string(some_value)`. on server side I have added validation to ensure frontend will only receive valid query strings, so we should be good here. Proof: [Click me](https://github.com/zhaoooyue/streamlit/blob/69f0673d6900bac0d6a30f07f62114ebbd5a6d23/lib/streamlit/__init__.py#L625-L630)